src/product_combination.rs

```diff
@@ -0,0 +1,80 @@
+/// An iterator to iterate through all product combinations of an iterator.
+///
+/// See [`.product_combination()`](../trait.Itertools.html#method.product_combination) for more information.
+pub struct ProductCombination<I: Iterator + Clone> {
+    data: I,
+    //Internal state is a vec of tuples containing an iterator and the last retrieved item from this iterator
+    internal_state: Vec<(I, I::Item)>
+}
+
+/// Create a new `ProductionCombination` from a clonable iterator.
+pub fn product_combination<I>(iter: I) -> ProductCombination<I>
+    where I: Iterator + Clone
+{
+    ProductCombination {
+        data: iter,
+        internal_state: Vec::new(),
+    }
+}
+
+impl<I: Iterator + Clone> ProductCombination<I> {
+    fn increment(&mut self) {
+        let mut pointer: usize = 0;
+        let mut tack_on = false;
+        loop {
+            match self.internal_state.get_mut(pointer) {
+                None => {
+                    tack_on = true;
+                    break;
+                },
+                Some(ref mut pair) => {
```

Comment:
We can't use default match binding in Itertools since it's still on Rust 1.12, but the tuple deref/item re-ref can still be done within the pattern match:
```rust
Some(&mut (ref mut iter, ref mut item)) => {
```